### PR TITLE
Define a reusable dataflow test task

### PR DESCRIPTION
**Pull request by AndrewShf**
Avoid duplication in similar gradle task, by reusing the configure and logic.

### CODE REVIEW

File: dataflow/build.gradle [hunk 1]
```diff
@@ -54,46 +55,61 @@ createDataflowShaded('shaded')
...
+        delete("tests/${dirName}/Test.class")
+        doLast {
+            javaexec {
+                workingDir = "tests/${dirName}"
+                if (!JavaVersion.current().java9Compatible) {
+                    jvmArgs += "-Xbootclasspath/p:${configurations.javacJar.asPath}".toString()
+                }
+                else if (JavaVersion.current() > JavaVersion.VERSION_11) {
+                    jvmArgs += [
+                            "--add-opens", "jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED",
```
Comment (zcai1) [inline dataflow/build.gradle:85]: nit: I think we should stick to 4 spaces for indentation?

File: dataflow/build.gradle [hunk 2]
```diff
@@ -54,46 +55,61 @@ createDataflowShaded('shaded')
...
-            executable 'diff'
-            args = ['-u', 'Expected.txt', 'Out.txt']
         }
     }
 }
 
+// run the liveVariableTest
+testDataflowAnalysisDemo("liveVariableTest", "live-variable", "livevar", "LiveVariable")
+// run the busyExpressionTest
+testDataflowAnalysisDemo("busyExpressionTest", "busy-expression", "busyexpr", "BusyExpression")
```
Comment (zcai1) [inline dataflow/build.gradle:111]: I think we don't have "busyExpressionTest" in this PR.

File: dataflow/build.gradle [hunk 3]
```diff
@@ -107,13 +123,13 @@ task issue3447Test(dependsOn: [assemble, compileTestJava], group: 'Verification'
...
             else if (JavaVersion.current() > JavaVersion.VERSION_11) {
                 jvmArgs += [
-                    "--add-opens", "jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED",
-                    "--add-opens", "jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED",
-                    "--add-opens", "jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED",
-                    "--add-opens", "jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED",
-                    "--add-opens", "jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED",
-                    "--add-opens", "jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED",
-                    "--add-opens", "jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED"
+                        "--add-opens", "jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED",
```
Comment (zcai1) [inline dataflow/build.gradle:126]: same comment about indentation here.

**Review by zcai1 — COMMENTED** [review]
Good job! LGTM with minor comments.

File: dataflow/build.gradle (hunk 2 above)
Comment (AndrewShf) [reply dataflow/build.gradle:111]: Thanks for pointing this out!

File: dataflow/build.gradle [hunk 4]
```diff
@@ -54,46 +55,59 @@ createDataflowShaded('shaded')
...
+                        "--add-opens", "jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED",
+                        "--add-opens", "jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED",
+                        "--add-opens", "jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED",
+                        "--add-opens", "jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED"
+                    ]
+                }
+
+                classpath = sourceSets.test.runtimeClasspath
+                classpath += sourceSets.test.output
+                mainClass = "${pkgName}.${className}"
```
Comment (wmdietl) [inline dataflow/build.gradle:97]: Is there a reason to split `pkgName` and `className`? Instead, could you just have a single class-name parameter that contains a fully-qualified name?

File: dataflow/build.gradle [hunk 5]
```diff
@@ -17,6 +17,7 @@ dependencies {
 
 // Shadowing Test Sources and Dependencies
 import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
+import org.gradle.plugin.use.internal.InjectedPluginClasspath
```
Comment (wmdietl) [inline dataflow/build.gradle:20]: I don't see a use of `InjectedPluginClasspath` anywhere below. Can you add a comment above this line saying what the import is for?

File: dataflow/build.gradle [hunk 6]
```diff
@@ -54,46 +55,59 @@ createDataflowShaded('shaded')
...
-            workingDir = 'tests/live-variable'
-            if (!JavaVersion.current().java9Compatible) {
-                jvmArgs += "-Xbootclasspath/p:${configurations.javacJar.asPath}".toString()
+/**
+ * create a task to run dataflow analysis demo
+ * @param taskName specify the taskName of created task
+ * @param dirName specify the directory name of the test file
+ * @param pkgName specify the package name of the demo
+ * @param className specify the class name of the demo
+ * @return
```
Comment (wmdietl) [inline dataflow/build.gradle:64]:
```suggestion
 * Create a task to run a dataflow analysis demo
 * @param taskName specify the taskName for the created task
 * @param dirName specify the directory name of the test file
 * @param pkgName specify the package name of the demo
 * @param className specify the class name of the demo
 * @return
```

File: dataflow/build.gradle [hunk 7]
```diff
@@ -54,46 +55,59 @@ createDataflowShaded('shaded')
...
+ */
+def testDataflowAnalysisDemo(taskName, dirName, pkgName, className) {
```
Comment (wmdietl) [inline dataflow/build.gradle:66]: Is the `Demo` necessary? The name already starts with `test`.

File: dataflow/build.gradle (hunk 4 above)
Comment (wmdietl) [reply dataflow/build.gradle:97]: Ah, you used `className` in the test description. My suggestion would just use the directory name there.

File: dataflow/build.gradle [hunk 8]
```diff
@@ -54,46 +55,59 @@ createDataflowShaded('shaded')
...
+    tasks.create(name: taskName, dependsOn: [assemble, compileTestJava], group: 'Verification') {
+        description "Test the ${className} analysis test for dataflow framework."
```
Comment (wmdietl) [inline dataflow/build.gradle:68]:
```suggestion
        description "Run the ${dirName} dataflow framework test."
```

File: dataflow/build.gradle [hunk 9]
```diff
@@ -54,46 +55,59 @@ createDataflowShaded('shaded')
...
-        }
-        exec {
-            workingDir = 'tests/live-variable'
-            executable 'diff'
-            args = ['-u', 'Expected.txt', 'Out.txt']
         }
     }
 }
 
+// run the liveVariableTest
```
Comment (wmdietl) [inline dataflow/build.gradle:108]:
```suggestion
```

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks! A few small comments & suggestions.

File: dataflow/build.gradle (hunk 5 above)
Comment (AndrewShf) [reply dataflow/build.gradle:20]: Ah, I even don't know when I added that, it is a mistake. I will delete this line.

File: dataflow/build.gradle (hunk 7 above)
Comment (AndrewShf) [reply dataflow/build.gradle:66]: Yes, it's a bit redundant. I will remove that.

File: dataflow/build.gradle (hunk 4 above)
Comment (AndrewShf) [reply dataflow/build.gradle:97]: Ok, now I use the full qualified class name.

File: dataflow/build.gradle [hunk 10]
```diff
@@ -54,46 +54,58 @@ createDataflowShaded('shaded')
...
-    doLast {
-        javaexec {
-            workingDir = 'tests/live-variable'
-            if (!JavaVersion.current().java9Compatible) {
-                jvmArgs += "-Xbootclasspath/p:${configurations.javacJar.asPath}".toString()
+/**
+ * Create a task to run a dataflow analysis demo
+ * @param taskName specify the taskName for the created task
+ * @param dirName specify the directory name of the test file
+ * @param pkgName specify the package name of the demo
```
Comment (wmdietl) [inline dataflow/build.gradle:61]:
```suggestion
```

File: dataflow/build.gradle [hunk 11]
```diff
@@ -54,46 +54,58 @@ createDataflowShaded('shaded')
...
+ * @param className specify the class name of the demo
```
Comment (wmdietl) [inline dataflow/build.gradle:62]:
```suggestion
 * @param className the fully-qualified name of the test class
```

File: dataflow/build.gradle [hunk 12]
```diff
@@ -54,46 +54,58 @@ createDataflowShaded('shaded')
...
-    delete('tests/live-variable/Test.class')
-    doLast {
-        javaexec {
-            workingDir = 'tests/live-variable'
-            if (!JavaVersion.current().java9Compatible) {
-                jvmArgs += "-Xbootclasspath/p:${configurations.javacJar.asPath}".toString()
+/**
+ * Create a task to run a dataflow analysis demo
+ * @param taskName specify the taskName for the created task
+ * @param dirName specify the directory name of the test file
```
Comment (wmdietl) [inline dataflow/build.gradle:60]:
```suggestion
 * @param taskName the taskName for the created task
 * @param dirName the directory name of the test file
```

**Review by wmdietl — APPROVED** [review]
Thanks!

**Pull request merged**